### ISSUE TITLE
ClaRa non-linear system Jacobian singular at initialization

### ISSUE THREAD

**Issue by AnHeuermann** [OpenModelica/OpenModelica#8459]
### Description

Some models in ClaRa_dev that are now passing Frontend and Backend are failing at simulation time, see https://libraries.openmodelica.org/branches/master/ClaRa_dev/ClaRa_dev.html.

Some of these failing models:

- ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_1ph_shell 
- ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_2ph_shell
- ClaRa.Components.HeatExchangers.Check.Test_HEXvle2vle_L3_1ph_BU_simple
- ClaRa.Components.HeatExchangers.Check.Test_HEXvle2vle_L3_1ph_kA
- ClaRa.Components.TurboMachines.Pumps.Check.TestPump_L2_OffDesign

KINSOL is reporting singular Jacobian rows and finally fails to solve the non-linear system.

### Steps to Reproduce

Simulate `ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_1ph_shell` with omc.

```bash
startTime=0
stopTime=3000
tolerance=1e-06
numberOfIntervals=5000
stepSize=0.6
Regular simulation: ./ClaRa_dev_ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_1ph_shell -abortSlowSimulation -alarm=480 -lv LOG_STATS
stdout            | info    | Using sparse solver kinsol for nonlinear system 0 (139),
|                 | |       | because density of 0.02 remains under threshold of 0.10.
stdout            | info    | Using sparse solver kinsol for nonlinear system 1 (355),
|                 | |       | because density of 0.06 remains under threshold of 0.10.
stdout            | info    | The maximum density for using sparse solvers can be specified
|                 | |       | using the runtime flag '<-nlssMaxDensity=value>'.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 4 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 7 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 8 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 9 singular. See LOG_NLS for more information.
stdout            | warning | ##KINSOL## Jacobian row 10 singular. See LOG_NLS for more information.
assert            | debug   | Solving non-linear system 355 failed at time=0.
|                 | |       | For more information please use -lv LOG_NLS.
assert            | error   | Failed to solve the initialization problem with global homotopy with equidistant step size.
assert            | debug   | Unable to solve initialization problem.
assert            | info    | simulation terminated by an assertion at initialization
```

### Version and OS

 - OpenModelica Version: OMCompiler v1.19.0-dev.545+g5c10d4c725
 - OS: Ubuntu 20.04.3 LTS
 - Versions of used Modelica libraries: ClaRa_dev @ c57bace0acbe86a8ae2ccf3e427c60093c5ac95c

**Comment by AlesVojacek** [comment]
- Dymola for this example (ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_1ph_shell) creates only 1 non-linear system (iteration variable: Volume.h) at initialization. OpenModelica creates non-linear system at initialization of size 11. (Iteration variable: volume.h, volume.p, volume.bulk.cp, volume.bulk.cv, volume.bulk.beta, volume.bulk.kappa, etc.), but the cp, cv, beta, etc. are calculated from the fluid object ("bulk") and not to vary those functions to fit.  
- This example (ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_1ph_shell) however runs through if the initOption=0, i.e. no initial equation is applied.

**Comment by casella** [comment]
I haven't looked into this in detail, but from the outside it looks like there is some issue with dependency analysis when fluid property records are involved. It seems that the backend thinks there are some dependencies (hence the large non-linear system), but at the end of the day at least some of them are not real, hence some rows of the Jacobian turn out to be zero at runtime, causing the solver to fail.

It could also be that this involves $cse variables containing the record output of fluid computation function calls, which are not handled correctly.

**Comment by AnHeuermann** [comment]
Using the internal difference quotient for Jacobian of the linear system and disabeling scaling when using KINSOL is working as well.
We are using the Jacobian to calculate scaling factors for F.

Maybe `nlsSparseJac` from kinsolSolver.c is wrong or the sparsity pattern is wrong.
It also seems there is an error in `nlsSparseSymJac`. I'm getting a segmentation fault when calling it (needs `--generateSymbolicjacobian`).

**Comment by AnHeuermann** [comment]
The sparsity pattern for the 11x11 Jacobian is very wrong. @kabdelhak suspects that this is because we have functions returning tuples inside the Jacobian.

From `-d=dumpLoops`
```
residual equations (4)
1/1 (1): valveLinear_1_XRG1.eye_int[1].m_flow = valveLinear_1_XRG1.pressureLoss.aperture_ * valveLinear_1_XRG1.pressureLoss.m_flow_nom * valveLinear_1_XRG1.pressureLoss.Delta_p / valveLinear_1_XRG1.pressureLoss.Delta_p_nom   [dynamic |0|0|0|0|]
2/2 (9): (Volume.heattransfer.fluid_wall.q, Volume.heattransfer.cp_w, Volume.heattransfer.fluid_wall.cv, Volume.heattransfer.fluid_wall.beta, Volume.heattransfer.fluid_wall.kappa, Volume.heattransfer.fluid_wall.drhodp_hxi, Volume.heattransfer.fluid_wall.drhodh_pxi, _, Volume.heattransfer.fluid_wall.w, Volume.heattransfer.fluid_wall.gamma) = TILMedia.Internals.VLEFluidObjectFunctions.additionalProperties_phxi(Volume.p, Volume.heattransfer.fluid_wall.h, {}, Volume.heattransfer.fluid_wall.vleFluidPointer)   [dynamic |0|0|0|0|]
3/11 (9): (Volume.bulk.q, Volume.bulk.cp, Volume.bulk.cv, Volume.bulk.beta, Volume.bulk.kappa, Volume.bulk.drhodp_hxi, Volume.bulk.drhodh_pxi, _, Volume.bulk.w, Volume.bulk.gamma) = TILMedia.Internals.VLEFluidObjectFunctions.additionalProperties_phxi(Volume.p, Volume.h, {}, Volume.bulk.vleFluidPointer)   [dynamic |0|0|0|0|]
4/20 (1): Volume.summary.outline.Q_flow_tot = homotopy(Volume.heattransfer.alpha * Volume.geo.A_heat_CF[1] * Volume.heattransfer.Delta_T_mean, Volume.heattransfer.alpha_nom * Volume.geo.A_heat_CF[1] * Volume.heattransfer.Delta_T_mean)   [dynamic |0|0|0|0|]
```

we can see, that there are a lot of residual equations are missing.

**Comment by AnHeuermann** [comment]
I've implemented a simple workaround for the time beeing. Test if the sparsity pattern has not enough non-zero elements. If so remove it fall back to the internal difference quotient for the Jacobian instead of the sparse one.

I'll open a PR. When @kabdelhak and @phannebohm fix the issue in the (new) backend we can remove the sanity check (or keep it).

**Comment by casella** [comment]
> I've implemented a simple workaround for the time beeing. Test if the sparsity pattern has not enough non-zero elements.

Do you mean, e.g., getting rows without any non-zero elements?

**Comment by AnHeuermann** [comment]
Ecactly. I'm testing if nnz (number of non-zeros) is larger than the size of the loop and if each row has at least one non-zero element.

**Comment by AnHeuermann** [comment]
I think our sparsity pattern is in a lot of cases very wrong, see [this PR run from Jenkins](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-8907/1/tests).  That would explain our problems when using kinsol for more systems / when chaning the default size when to switch from a dense solver to a sparse one.

**Comment by casella** [comment]
> I think our sparsity pattern is in a lot of cases very wrong, see [this PR run from Jenkins](https://test.openmodelica.org/jenkins/blue/organizations/jenkins/OpenModelica/detail/PR-8907/1/tests). That would explain our problems when using kinsol for more systems / when chaning the default size when to switch from a dense solver to a sparse one.

@AnHeuermann I'm a bit confused. I understood PR #8970 contained workarounds to make broken examples to work, not the other way round...

**Comment by perost** [comment]
> The base problem of the solver failing seems to be resolved with #8913 but there are further problems.
> 
> From flat modelica i can see that our frontend creates the following function call:
> 
> ```
>   massFlowSource.fluidOut.h = TILMedia.Internals.VLEFluidConfigurations.FullyMixtureCompatible.VLEFluidObjectFunctions.specificEnthalpy_pTxi(massFlowSource.fluidOut.p, massFlowSource.fluidOut.T, {}, massFlowSource.fluidOut.vleFluidPointer);
> ```
> 
> The crucial part is the empty list representing the mass fraction `xi`. This function fails later on during simulation and our assumption is that it is due to this empty list.
> 
> So the basic questions are: Is this list supposed to be empty? Is the function capable of handling empty lists? Or is there maybe any other reason that this function could fail?

The reason for the `{}` is that `xi` is an empty array (at least that's what the NF thinks), and as such doesn't exist in the flat model. So we replace any uses of it with an empty array expression to avoid having references to non-existent variables which causes compilation issues.

**Comment by casella** [comment]
I am not an expert in the TILMedia package (so I'd also add @AlesVojacek to the loop) but I can't see why having empty `xi` would be an error. `xi` is the array of independent mass fractions in the fluid. If you have a pure or pseudo-pure fluid, the size of that array is zero, and this should be fully supported by the toolchain

**Comment by casella** [comment]
> I am not an expert in the TILMedia package (so I'd also add @AlesVojacek to the loop) but I can't see why having empty `xi` would be an error. `xi` is the array of independent mass fractions in the fluid. If you have a pure or pseudo-pure fluid, the size of that array is zero, and this should be fully supported by the toolchain

Of course the external functions need to manage that, but that's none of our business, as long as the Jacobian sparsity pattern detection is concerned.

**Comment by kabdelhak** [comment]
Maybe i am just doing something wrong on my machine and don't have it properly set up. Can someone confirm with the current master?

**Comment by AnHeuermann** [comment]
The first example `ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_1ph_shell` is simulating for me with `omc v1.20.0-dev-41-g8529da7901` and the solution is correct (according to diffSimulationResults).

**Comment by AnHeuermann** [comment]
The other examples are failing at initialization:

```
messages = "Simulation execution failed for model: ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_2ph_shell
assert            | debug   | Solving non-linear system 131 failed at time=0.
|                 | |       | For more information please use -lv LOG_NLS.
assert            | error   | Failed to solve the initialization problem with global homotopy with equidistant step size.
assert            | debug   | Unable to solve initialization problem.
assert            | info    | simulation terminated by an assertion at initialization
```
[Test_HEXvle2vle_L3_1ph_BU_simple_res.log](https://github.com/OpenModelica/OpenModelica/files/8660562/Test_HEXvle2vle_L3_1ph_BU_simple_res.log)
[Test_HEXvle2vle_L3_1ph_kA.log](https://github.com/OpenModelica/OpenModelica/files/8660563/Test_HEXvle2vle_L3_1ph_kA.log)
[TestPump_L2_OffDesign.log](https://github.com/OpenModelica/OpenModelica/files/8660564/TestPump_L2_OffDesign.log)
[Validation_VolumeVLE_L2_HeatTransfer_1ph_shell.log](https://github.com/OpenModelica/OpenModelica/files/8660565/Validation_VolumeVLE_L2_HeatTransfer_1ph_shell.log)
[Validation_VolumeVLE_L2_HeatTransfer_2ph_shell.log](https://github.com/OpenModelica/OpenModelica/files/8660566/Validation_VolumeVLE_L2_HeatTransfer_2ph_shell.log)

**Comment by AnHeuermann** [comment]
@kabdelhak and myself tried to figure out why `ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_2ph_shell` isn't working.

In non-linear system 131 we have equations involving `homotopy`, `noEvent`, `actualStream` and `inStream`.
The replacement for `actualStream` looks good. We can't tell if `inStream` is replaced correctly.

When using only the simplified version from the homotopy equation (`--replaceHomotopy=simplified`) the system can be solved.
Using only the actual one (`--replaceHomotopy=actual`) it can't be solved.

Here are the non-linear systems:
[simplified.log](https://github.com/OpenModelica/OpenModelica/files/8670962/simplified.log)
[actual.log](https://github.com/OpenModelica/OpenModelica/files/8670963/actual.log)

**Comment by casella** [comment]
If the simplified system can be solved, then the homotopy transformation should start, and eventually fail for some value of lambda, since the system is not solved, eventually. Have you tried to log the homotopy transformation to see what happens, and possibly to change the number of steps or switch to the adaptive homotopy solver?

**Comment by kabdelhak** [comment]
> If the simplified system can be solved, then the homotopy transformation should start and be successful at lambda = 0, where the symbolic substitution homotopy with simplified is exploited, then eventually fail for some value of lambda > 0, since the system is not solved, eventually. Have you tried to log the homotopy transformation to see what happens, and possibly to change the number of steps or switch to the adaptive homotopy solver?

Yea we changed the number of homotopy steps and we raised the number of steps that the solver is allowed to take. It always resulted in convergence up until just before `lambda = 1`. So every step before that was fine, but the last one seemed to be singular. We looked into the definition on `stream` and the corresponding functions `actualStream()` and `inStream()` and noticed some things that might go wrong:
1. the stream connection has singularities around zero flow - might this be the case here? how would we avoid that?
2. `actualStream()` and `inStream()` have to be resolved in a very specific way (replaced by the frontend), maybe we are doing something incorrectly there? From what i can see `inStream()` has to sum up other `stream` variables from connectors connected to the connector of which the `stream` variable was used in `inStream()`. I did not really understand the specification here to be honest.
3. The initialization system seems to be solvable for any step before `lambda=1` so it has to be singular numerically and not structurally, so my guess is that the jacobian is at least solvable (maybe still wrong). Maybe there is something wrong with the process of numerical solving? Some kind of stable fix point iteration where we follow the same trajectory over and over again?

**Comment by phannebohm** [comment]
> 3. The initialization system seems to be solvable for any step before `lambda=1` so it has to be singular numerically and not structurally, so my guess is that the jacobian is at least solvable (maybe still wrong). Maybe there is something wrong with the process of numerical solving? Some kind of stable fix point iteration where we follow the same trajectory over and over again?

Is it not possible for the system to be structurally singular? I don't see why it can't. The homotopy expressions might hide the singularity until `lambda=1`. That would also fit to your observation 1 and Andreas' comment:

> When using only the simplified version from the homotopy equation (--replaceHomotopy=simplified) the system can be solved. Using only the actual one (--replaceHomotopy=actual) it can't be solved.

I think we should investigate this. If the system is singular, this would be another reason to introduce a lambda_1 system, analogous to the lambda_0 system. Then we could find structural singularities there.

**Comment by casella** [comment]
> I think we should investigate this.

I'll do that asap.

> If the system is singular, this would be another reason to introduce a lambda_1 system, analogous to the lambda_0 system. Then we could find structural singularities there.

True, but that would make code generation even slower than it already is.

**Comment by casella** [comment]
I wonder if the reason why the system fails at lambda = 1 is because of a bad tearing that makes the system singular. Unfortunately, I'm not sure turning off tearing completely is not going to break the initialization elsewhere, what do you think?

**Comment by anotheruserofgithub** [comment]
@casella I don't know if this can help, but for a while now I always use these compiler and simulation settings for big multi-body models:
```modelica
annotation(
  experiment(StartTime = 0, StopTime = 1, Tolerance = 1e-6, Interval = 0.002),
  __OpenModelica_commandLineOptions = "-d=aliasConflicts,initialization,stateselection,tearingdump,LSanalyticJacobian,NLSanalyticJacobian,forceNLSanalyticJacobian --matchingAlgorithm=PFPlusExt --indexReductionMethod=dynamicStateSelection --daeMode=false --generateSymbolicJacobian=true --tearingStrictness=veryStrict --maxSizeLinearTearing=1000000 --maxSizeNonlinearTearing=1000000 --recursiveTearing=0 --initOptModules+=recursiveTearing --postOptModules+=recursiveTearing",
  __OpenModelica_simulationFlags(s = "dassl", jacobian = "symbolical", noHomotopyOnFirstTry = "()", lssMaxDensity = "0.000001", nlssMaxDensity = "0.000001", lssMinSize = "1000000", nlssMinSize = "1000000", lv = "LOG_DT,LOG_DSS,LOG_LS,LOG_NLS,LOG_INIT,LOG_SOTI,LOG_SOLVER,LOG_EVENTS,LOG_STATS,LOG_DASSL,LOG_DASSL_STATES", ls = "totalpivot"));
```

I had severe problems (such as the ones you were commenting to yesterday, i.e. division by zero due to bad tearing, singular at initialization, sparse solvers failing, ...) with the model going bigger and more complicated, but six months ago I finally found those settings were doing the trick very well so I stick to them now.

Basically I force dense solvers and very strict tearing. Recursive tearing was not always successful so I kept it disabled.

Hope this can serve someone else, and help you find a more generic and transparent solution.

**Comment by casella** [comment]
@anotheruserofgithub, every user has its magic mix (or mixes) of flags that work for him/her, but that very much depends on the set of models that you are running. Also, over time some of these flags become obsolete at some point. More specifically
- I would not recommend noHomotopyOnFirstTry, in my experience if homotopy is declared in the model, it should be used
- the automatic selection of sparse linear solver was fixed for good in 1.19.0, from that version I believe it is no longer necessary to set lssMin/Max density and lssMin/Max size, just leave the choice to the tool. Unless you know exactly what you are doing.
- daeMode will never be set on by default, so there is no need to disable it; however, for certain models (e.g. 3-phase electro-magnetic circuit models) it can be very advantageous

**Comment by phannebohm** [comment]
As of OMCompiler v1.22.0-dev.245+g6a88ed1a95 this issue does not seem to be present anymore.

The following models still have nonlinear systems failing at initialization but don't report singular Jacobians:
- ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_2ph_shell
- ClaRa.Components.HeatExchangers.Check.Test_HEXvle2vle_L3_1ph_kA

The following models simulate now but suffer from numerical noise:
- ClaRa.Basics.ControlVolumes.FluidVolumes.Check.Validation_VolumeVLE_L2_HeatTransfer_1ph_shell
- ClaRa.Components.HeatExchangers.Check.Test_HEXvle2vle_L3_1ph_BU_simple
- ClaRa.Components.TurboMachines.Pumps.Check.TestPump_L2_OffDesign

Closing this in favor of #10980.
